### PR TITLE
Fix the BIO_addr test

### LINKED ISSUE #22218: 3.2-alpha2 test_bio_addr fails on NonStop
The `61-test_bio_addr` fails on NonStop x86 with reseeding not being detected. We are using the hardware randomizer on x86. The hardware randomizer has worked since 3.0, but this looks like it might be a new situation.

Assistance requested to localize what is or is not happening.

```
61-test_bio_addr.t ...................... 
# The results of this test will end up in test-runs/test_bio_addr
1..1
    # Subtest: ../../test/bio_addr_test
    1..1
        # Subtest: test_bio_addr_copy_dup
        1..6
        ok 1 - iteration 1
        ok 2 - iteration 2
        ok 3 - iteration 3
        ok 4 - iteration 4
        # ERROR: (bool) 'bio_addr_is_eq(src, dst) == true' failed @ /home/jenkinsbuild/.jenkins/workspace/OpenSSL-3.2_Monitor/test/bio_addr_test.c:145
        # false
        # OPENSSL_TEST_RAND_SEED=1695973622
        not ok 5 - iteration 5
        # ERROR: (bool) 'bio_addr_is_eq(src, dst) == true' failed @ /home/jenkinsbuild/.jenkins/workspace/OpenSSL-3.2_Monitor/test/bio_addr_test.c:145
        # false
        # OPENSSL_TEST_RAND_SEED=1695973622
        not ok 6 - iteration 6
    # OPENSSL_TEST_RAND_SEED=1695973622
    not ok 1 - test_bio_addr_copy_dup
../../util/wrap.pl ../../test/bio_addr_test => 255
not ok 1 - running bio_addr_test

#   Failed test 'running bio_addr_test'
#   at test/recipes/61-test_bio_addr.t line 20.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
```

**Configuration:**
```
Command line (with current working directory = .):

    perl ./Configure nonstop-nsx_64 --with-rand-seed=rdcpu enable-fips

Perl information:

    perl
    5.30.3 for NSV-D-nonstop_kernel

Enabled features:

    acvp-tests
    apps
    argon2
    aria
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    docs
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    ecx
    egd
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    http
    idea
    legacy
    loadereng
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    quic
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm2-precomp
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    asm                 [no asm_arch]    OPENSSL_NO_ASM
    brotli              [default]        OPENSSL_NO_BROTLI
    brotli-dynamic      [default]        OPENSSL_NO_BROTLI_DYNAMIC
    buildtest-c++       [default]
    winstore            [not-windows]    OPENSSL_NO_WINSTORE
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    default-thread-pool [cascade]        OPENSSL_NO_DEFAULT_THREAD_POOL
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    tfo                 [default]        OPENSSL_NO_TFO
    thread-pool         [cascade]        OPENSSL_NO_THREAD_POOL
    threads             [config]
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]        OPENSSL_NO_ZLIB
    zlib-dynamic        [default]        OPENSSL_NO_ZLIB_DYNAMIC
    zstd                [default]        OPENSSL_NO_ZSTD
    zstd-dynamic        [default]        OPENSSL_NO_ZSTD_DYNAMIC
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "c99",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-g -O2 -Wextensions -Wnowarn=203,220,272,734,770,1506 -Wbuild_neutral_library -Wverbose -Wtarget=tns/x -Wsystype=oss -Wlp64 -WIEEE_float",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "_TANDEM_SOURCE", "__NSK_OPTIONAL_TYPES__", "B_ENDIAN" ],
    disable => [ "threads" ],
    dso_scheme => "DLFCN",
    enable => [ "egd" ],
    ex_libs => "-lrld",
    includes => [  ],
    lflags => "-Wxld=\"-set systype oss\" -Wxld=\"-set data_model lp64\" -Wxld=\"-set floattype ieee_float\"",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "-Wshared",
    multibin => "64",
    multilib => "64",
    perl => "/usr/bin/perl",
    perl_platform => "Unix",
    shared_argfileflag => "-Wxld_obey=",
    shared_cflag => "",
    shared_defflag => "-Wxld_obey=",
    shared_defines => [  ],
    shared_extension => ".so",
    shared_ldflag => "-Wshared",
    shared_rcflag => "",
    shared_target => "nonstop-shared",
    sys_id => "TANDEM",
    thread_defines => [  ],
    thread_scheme => "(unknown)",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    ASFLAGS         =
    CC              = c99
    CFLAGS          =
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

### PR DESCRIPTION
**Pull request by mattcaswell**
The BIO_addr test is failing on non-stop. The length of the data is larger than the size we have allocated for it. We dynamically allocate instead.

Fixes #22218

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by rsbeckerca** [comment]
LGTM

**Review by paulidale — APPROVED** [review]

**Comment by nhorman** [comment]
I'm sorry, I'm not sure I see how this works (though I might be missing something).  BIO_ADDR_rawaddress only returns 3 length sizes, sizeof(s_in.sin_addr) (4 bytes), sizeof (_sin6.sin6_addr)  (16 bytes) or strlen(s_un.sun_path) (108 bytes max),  and sockaddr storage is large enough to hold all of those.  How does dynamically allocating a buffer that will be one of those three sizes, rather than using an automatically allocated buffer avoid the problem nonstop was seeing?

**Comment by rsbeckerca** [comment]
> I'm sorry, I'm not sure I see how this works (though I might be missing something). BIO_ADDR_rawaddress only returns 3 length sizes, sizeof(s_in.sin_addr) (4 bytes), sizeof (_sin6.sin6_addr) (16 bytes) or strlen(s_un.sun_path) (108 bytes max), and sockaddr storage is large enough to hold all of those. How does dynamically allocating a buffer that will be one of those three sizes, rather than using an automatically allocated buffer avoid the problem nonstop was seeing?

@nhorman Please examine the diagnostics for #22218 for why the code prior to this PR does not work.

**Comment by nhorman** [comment]
@rsbeckerca I have, and I don't think the analysis there really answers my question.  The analysis seems to center around the notion that nonstop unix addresses are "very large" (ostensibly implying addresses larger than 108 bytes).  That might be the case, but unix addresses as I read them are limited to 108 bytes (as defined in the sockaddr_storage struct).  If non stop is exceeding that size in the bio_addr, that seems like a non stop issue to fix, not something we should paper over here.  I could well be missing something, but I'm not seeing it

**Comment by rsbeckerca** [comment]
> @rsbeckerca I have, and I don't think the analysis there really answers my question. The analysis seems to center around the notion that nonstop unix addresses are "very large" (ostensibly implying addresses larger than 108 bytes). That might be the case, but unix addresses as I read them are limited to 108 bytes (as defined in the sockaddr_storage struct). If non stop is exceeding that size in the bio_addr, that seems like a non stop issue to fix, not something we should paper over here. I could well be missing something, but I'm not seeing it

I am very happy to test out whatever suggested change you have to fix this easily demonstrated defect in the code. That fact that NonStop has a different data structure is not my doing, but the data structure itself did cause a breakage in the test and is is pointing out an edge condition that the existing code does not support. Please reference your pull request in this thread.

**Comment by nhorman** [comment]
@rsbeckerca I'm really just trying to understand how this fixes the problem as described, As I understand @22218, Nonstop uses AF_UNIX for some purpose, and in some way overloads the size of the AF_UNIX socket structure to point to data greater than 108 bytes.  Do I have that correct?  If so, then BIO_ADDR_rawaddress will call strlen to get the size of the buffer (which may well return a larger value.  However, sockaddr_un in the unix case is still defined as:
```
struct sockaddr_un {
    short sun_family;           /* AF_UNIX */
    char sun_path[108];         /* path name (gag) */
};
```
so if on nonstop systems, strlen(ap->s_un.sun_path) returns a value greater than 107 (accounting for the NULL terminator), then whatever data the nonstop platform has stored there seems to me, by definition to have written past its allocated memory, which would be the real problem here, that this fix may just be covering up.

Again, I don't have access to a nonstop system at the moment, so I may be missing something here (perhaps they've modified the definition of sockaddr_un to be larger), I'm just trying to get clarity on how this works.

**Comment by nhorman** [comment]
It just occured to me what might be happening here.  If, on nonstop platforms, AF_UNIX addresses are using  the complete address space in s_un.sun_path (i.e 108 bytes) without a NULL terminator, then strlen on that address would potentially return a very large value (encompassing whatever size is found at the first NULL terminator it finds in ram, which will be at some point beyond the sockaddr_storage allocated memory.  Thats bad, as the memory will be whatever garbage is in there.  If thats what they are doing, then it would seem to me that the more correct fix for this issue would be something like this:
https://github.com/nhorman/openssl/commit/42f9e904504ea8e9b1649596e7a7cae3cdff8c71

@rsbeckerca could you please try testing with that fix to see what the result is?  Thank you!

**Comment by nhorman** [comment]
please see comment https://github.com/openssl/openssl/issues/22218#issuecomment-1752097842
this fix shouldn't be merged, as the issue appears to be in the nonstop platforms definition of sockaddr_storage

**Comment by rsbeckerca** [comment]
> please see comment [#22218 (comment)](https://github.com/openssl/openssl/issues/22218#issuecomment-1752097842) this fix shouldn't be merged, as the issue appears to be in the nonstop platforms definition of sockaddr_storage

I do not agree. Proposing a solution requiring an operating system change is unacceptable.

**Comment by nhorman** [comment]
but the break is clearly within the operating system.  We might consider a workaround, but at one point or another the only way to fix this properly is to get the NonStop os to issue a fix.

**Comment by rsbeckerca** [comment]
> but the break is clearly within the operating system. We might consider a workaround, but at one point or another the only way to fix this properly is to get the NonStop os to issue a fix.

I do not understand what you mean here. What break specifically? The size of sun_part? And a fix to the OS will take 5-7 years as previously stated, by their support policy, even if the fix was made last January, I still have to support OpenSSL for existing customers who use older versions of the operating system - unless I resign from my position, which this whole situation is making me seriously consider.

**Comment by nhorman** [comment]
Apologies, let me elaborate (and I'm sorry If I'm covering obvious ground, I just want to be complete):

POSIX defines structures to manage sockets for every supported address family that the enumerate:
* [sockaddr_in](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/netinet/in.h.html) for AF_INET sockets
* [sockaddr_in6](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/netinet/in.h.html) for AF_INET6 sockets
* [sockaddr_un](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/un.h.html) for AF_UNIX sockets

There are more, but for the purpose of this discussion, that's sufficient

Each of those structures contains some common data (the family type designator), followed by family specific data (4 bytes in sockaddr_in to hold an ipv4 address, 16 bytes in sockaddr_in6 to hold an ipv6 address, and a platform specific number of bytes to hold the sun_path address for sockaddr_un structs).  In that last case, linux  and most unixes define that length to be 128 bytes, though posix doesn't require that, it can be any value, and NonStop seems to take advantage of that by defining the length to be 1024 bytes in that path, which is fine.

These definitions come with a programming problem however, manipulating sockets in application code implies that the application code in question already know the family type, so the appropriate struct can be used.  It would be nice if application developers could write common code with a single struct that can manage sockets generically, without having to know what kind of socket structure to use, and when the family is known, or becomes relevant to the application, said single struct could be cast to the appropriate family type without having to do any reallocations or complex copying.

POSIX obliges here by defining another struct, [sockaddr_storage](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html), which is made available for exactly the above purpose.  That is to say, an application that whishes to generically manipulate sockets without having to use family-specific structs, can instead allocate a sockaddr_storage struct and use that in place of sockaddr_in/sockaddr_in6/sockaddr_un/etc safely.  This is exactly what we do in our test_bio_addr_copy_dup test.  We create sockets of multiple different types, clone/copy them, and compare them for equality.  In the bio_addr_is_eq function, instead of allocating a sockaddr of every different family, we allocate sockaddr_storage structs instead, relying on the guarantee that that struct is large and aligned enough to hold any type of socket family struct.

POSIX makes this guarantee for sockaddr_storage by requiring that (as noted [here](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html)) that the implementing operating system define the struct to conform to two requirements:
1) It has to be large enough to hold any type of socket struct (sockaddr_in/sockaddr_in6/sockaddr_un/etc)
2) It has to be aligned at appropriate boundaries such that sockaddr_storage can be cast to any of the above families, and have that family's internal pointers access that cast structs data fields properly

This is where the problem happens. NonStop, when it implemented sockaddr_un, chose to increase the size of sun_path to 1024 bytes.  That's perfectly valid under POSIX.  However, it failed to correspondingly increase the size of sockaddr_storage to accommodate that larger sun_path array in sockaddr_un, violating the first requirement that POSIX holds the implementation to.  

As a result, because the bio_addr test in question creates an address of the defined size of sun_path (1024 bytes), and then uses sockaddr_storage in its check for equality (based on the now-incorrect reliance on the fact that sockaddr_storage is large enough to hold a sockaddr_un struct), and so winds up writing past its allocated memory, corrupting the next item on the stack.

So, in summary, the issue here is that NonStop didn't fulfill its promise when implementing sockaddr_storage to make it large enough to hold a sockaddr of any type (specifically sockaddr_un), and the result is that applications relying on that promise break, by writing to memory that isn't allocated.

Ironically, on PS-RISC systems where the stack grows upwards, this might have been considered a security bug, as attackers could have used this to overwrite the return address of a function on the stack (though I think all nonstop systems are ia64/x86_64, so thats probably not an issue)

**Comment by levitte** [comment]
> I'm sorry, I'm not sure I see how this works (though I might be missing something). BIO_ADDR_rawaddress only returns 3 length sizes, sizeof(s_in.sin_addr) (4 bytes), sizeof (_sin6.sin6_addr) (16 bytes) or strlen(s_un.sun_path) (108 bytes max), and sockaddr storage is large enough to hold all of those. How does dynamically allocating a buffer that will be one of those three sizes, rather than using an automatically allocated buffer avoid the problem nonstop was seeing?

Quite simple: the length there relies on the actual sizes of the sockaddr structures that BIO_ADDR supports, or in the AF_UNIX case, on the length of the path.  That's perfectly legitimate to do.

Also, there are platforms that do not have `sockaddr_storage` at all.  VMS version 8 is one of those.  So I'd say that avoiding to use that structure at all is a win for now.

**Comment by t8m** [comment]
@nhorman I disagree that we should add this workaround only if the NonStop team acknowledges they have a bug in their POSIX conformance. I do not think we should held users of OpenSSL hostage even if they are on a _broken_ platform. There are plenty of other workarounds for OS issues in the OpenSSL code and we never made the requirement to have these issues being acknowledged by the respective OS vendors.

If this was code in other areas than a testcase I'd say this workaround should be made conditional only for NonStop, but doing this just for a test code seems unnecessarily complicating the implementation. So IMO this is OK as is.

**Comment by mattcaswell** [comment]
> As a result, because the bio_addr test in question creates an address of the defined size of sun_path (1024 bytes), and then uses sockaddr_storage in its check for equality (based on the now-incorrect reliance on the fact that sockaddr_storage is large enough to hold a sockaddr_un struct),

Actually what BIO_ADDR_rawaddress returns is either the `sin_addr` field, or the `sin6_addr` field, or the `sun_path` field as appropriate, i.e. it doesn't return `sockaddr_un`, it returns `sockaddr_un.sun_path`. So the original test's use of `socaddr_storage` is not actually correct (you cannot cast the return value of `BIO_ADDR_rawaddress` to `sockaddr_un`). Although it should still be big enough (actually slightly bigger than it needs to be). But this type confusion in the original test is a little odd, and removing it is a good thing IMO.

> and so winds up writing past its allocated memory, corrupting the next item on the stack.

To be strictly accurate this never actually happened since the test checked that the sizeof was big enough before actually copying any data around. This check was why the test was failing.

> @nhorman I disagree that we should add this workaround only if the NonStop team acknowledges they have a bug in their POSIX conformance. I do not think we should held users of OpenSSL hostage even if they are on a broken platform. There are plenty of other workarounds for OS issues in the OpenSSL code and we never made the requirement to have these issues being acknowledged by the respective OS vendors.

Yes, I agree with @t8m. It is probably a good idea for someone to file a bug with HP. But we have many workarounds for dubious platform behaviour (or just plain platform bugs) in OpenSSL.

Ultimately this is just a minor workaround in a test case. Given the type confusion in the original test I don't see a problem with doing things this way.

> If this was code in other areas than a testcase I'd say this workaround should be made conditional only for NonStop, but doing this just for a test code seems unnecessarily complicating the implementation

Yes.

**Comment by rsbeckerca** [comment]
@mattcaswell I am going to work on a case with HPE on this, but a fix, if there is one, will take time. Do you want a discussion on this or should I just post to the mailing list?

**Comment by mattcaswell** [comment]
Feel free to ping me directly via email if you want to work out the specifics of the report to HPE.

**Comment by rsbeckerca** [comment]
@mattcaswell I opened up my profile temporarily. Please send me an email.
